### DISCUSSION TITLE
Request for documentation on argsToTemplate for Angular Storybook

### DISCUSSION THREAD

**Discussion by shellwe** [storybookjs/storybook#28243] [Help]
### Summary

So on this page there is a reference and a single sentence describing argsToTemplate
https://storybook.js.org/tutorials/intro-to-storybook/angular/en/simple-component/

But there is no mention anywhere else how it works and how it is better than manually placing variables. It just says it makes event binding far easier but doesn't describe what you need to do, it's just in the example.

I don't know if this is the best place to ask, but some additional documentation on argsToTemplate would be greatly appreciated.

**Comment by smlimasu** [comment]
See: https://storybook.js.org/docs/api/csf#custom-render-functions 

My personal understanding: 

In cases where you need to provide a render template (if its just your components args - see "CSF 3 - default render function" you do not need to set the render template string) - you can use the argsToTemplate Function to spread the args that you've set in the story or the (meta.args!) to the components string template. You could also use the options to include or exclude some args. 

It's very convenient when you need to have the string template, but don't want to set all inputs and outputs manually all the time. 

Example from the Documentation: 
```
export const Text: Story = {
  render: (args) => ({
    props: args,
    // The argsToTemplate helper function converts the args to property and event bindings.
    // You could also write the template in plain HTML and bind to the component's inputs and outputs yourself:
    // <storybook-button ["label"]="label" (onClick)="onClick($event)">
    // We don't recommend the latter since it can conflict with how Storybook applies arguments via its controls addon.
    // Binding to the component's inputs and outputs yourself will conflict with default values set inside the component's class.
    // In edge-case scenarios, you may need to define the template yourself, though.
    template: `<storybook-button ${argsToTemplate(args)}></storybook-button>`,
  }),
  args: {
    label: 'Hello',
    onClick: action('clicked'),
  },
}; 
```

With a subset of args - only conceptual example i just drafted, there are other or maybe better ways to spread args, its just to show argsToTemplate in the template string: 

```
export const Button: Story<StorybookButtonComponent> = {

args: {
label: 'firstButton',
},
  render: (args) => ({
    props: args,
    label2: 'ThirdButton',

    template: `
    //the first button will have the label "firstButton" and all other components args from meta or components default inferred
    <storybook-button ${argsToTemplate(args)}></storybook-button>
    //hypothetical example where you need to define a custom template like a second button or some other components inside your components template that are not defined via args (f.ex. ng-container?)
     <storybook-button ${argsToTemplate(args)}><icon [name]="info"></icon>SomeRandomText</storybook-button>
    <storybook-button ${argsToTemplate(args, { exclude: ['label'] })}  ["label"]="SecondButton"></storybook-button>
    //will show "ThirdButton" as the label.
        <storybook-button ${argsToTemplate(args, { exclude: ['label'] })}  ["label"]="label2"></storybook-button>
    `,
  }),
};
```

So now, with args to template, storybook can populate your components args in a pretty way to the "show code" section and to the controls addon. And you save yourself some manual work. But as the documentation mentioned, the best way is to try to not use custom templates when the render function is unneccessary to convert your args to the components default template. -> https://storybook.js.org/docs/api/csf#default-render-functions 

(I hope what I've just explained is okay, please correct me or show more examples (maybe how to spread other stories args?), storybooks documentation + angular is so confusing sometimes,  so that sometimes I'm just guessing with try + error (: )

**Reply by Marklb** [reply]
This looks correct in how it is described. The following is minor corrections that I think your second example would need. You set `label2` on the story return object, not `props`. In the template, I see a `SecondButton` variable, but I don't see where it is from. So, the way `SecondButton` is used is correct, since it could be from `Meta` or a decorator, but I assumed you meant to use it as a string. The way you defined the inputs are `["label"]=`, but it should be `[label]=`.
```ts
export const Button: Story<StorybookButtonComponent> = {
  args: {
    label: 'firstButton',
  },
  render: (args) => ({
    props: {
      ...args,
      label2: 'ThirdButton',
    },
    template: `
      //the first button will have the label "firstButton" and all other components args from meta or components default inferred
      <storybook-button ${argsToTemplate(args)}></storybook-button>
      //hypothetical example where you need to define a custom template like a second button or some other components inside your components template that are not defined via args (f.ex. ng-container?)
      <storybook-button ${argsToTemplate(args)}><icon [name]="info"></icon>SomeRandomText</storybook-button>
      <storybook-button ${argsToTemplate(args, { exclude: ['label'] })}  label="SecondButton"></storybook-button>
      //will show "ThirdButton" as the label.
      <storybook-button ${argsToTemplate(args, { exclude: ['label'] })}  [label]="label2"></storybook-button>
    `,
  }),
};
```

**Reply by smlimasu** [reply]
You are absolutely correct Mark, sorry for the typos, yes the "secondButton" was meant as a string and the inputs should be [label]. 

If you don't bother asking me another question I just can't find any current examples how to reuse stories without mapping all the inputs manually again.

So let's say the second and third button are another story in that file and we set all the button args in args there. 

Now I wanna show a composed story of all 3 Buttons at once, reusing these stories (or even better also with their render functions when I needed to pass a custom template in the stories before). So that we can see 3 buttons underneath each other in the most efficient manner and that the show code section will in the best case show the exact input values for each button?

Like (conceptually) argsToTemplate(secondButtonStory.args)? Maybe? 

I can only get it to work with meta.args.

I would be so glad if you could show how you would do that? Or with a decorator? What's your best practice way to show multiple, but same components? Thanks in advance!
